### ISSUE TITLE
GPU parity FAILED: cosine=-0.005 for Qwen2.5-7B GQA (hidden=3584, heads=28, kv=4) on sm_121

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#559]
## Root Cause (Five-Whys)

**CORRECTNESS-011 layer trace output:**
```
PARITY-GATE FAILED: GPU computes a DIFFERENT function than CPU.
Cosine similarity: -0.005182 (required: ≥0.98)
CPU argmax: 334 | GPU argmax: 8127
Max absolute logit difference: 19.5080
```

1. Why GPU ≠ CPU? Cosine similarity -0.005 (completely uncorrelated, not FP rounding)
2. Why completely wrong? GPU forward pass computes an entirely different function
3. Why different function? Model dimensions `hidden=3584, heads=28, kv_heads=4` (GQA)
4. Which kernel? GQA attention with 28:4 head ratio (7 query heads per KV head)
5. Root cause: GQA CUDA kernel handles non-power-of-2 head ratios incorrectly

## Diagnosis

- This is NOT FP rounding (cosine would be ~0.999)
- This is NOT a driver issue (driver 590 same result as 580)
- This is a **logic bug** in the GQA attention kernel for head_ratio=7
- The cosine of -0.005 means GPU output is essentially random relative to CPU

## Model Dimensions

- hidden_dim: 3584
- num_heads: 28
- num_kv_heads: 4
- head_dim: 128 (3584/28)
- head_ratio: 7 (28/4) — non-power-of-2

## Hardware

NVIDIA GB10 (Blackwell sm_121), driver 590.48.01, CUDA 13.1

## Provable Contract

`ptx-target-parity-v1.yaml` equation `target_parity` is violated.
`gpu-context-health-v1.yaml` — the GPU produces wrong results, not just context poisoning.

## Fix Required

Fix the GQA attention kernel in trueno-gpu for head_ratio=7 on sm_121.
Verify with CORRECTNESS-011 layer trace: cosine must be ≥0.98.

**Comment by noahgift** [comment]
## Confirmed: RMSNorm KERNEL computes wrong sum-of-squares on sm_121

### Proof

Same input data (embedding), host vs GPU kernel:
- **Host sum_sq** (from GPU buffer, computed on CPU): 1.1233730316
- **Host RMS**: 0.0177324880  
- **GPU RMSNorm output** first 3: [-0.20145158, -0.10565072, 0.27226985]
- **CPU RMSNorm output** first 3: [-0.20431474, -0.10715230, 0.27613953]

The GPU input data is BYTE-IDENTICAL to CPU (verified by downloading from GPU and computing sq_sum on host). But the GPU RMSNorm kernel produces a 1.4% wrong normalization factor.

The `PreciseRmsNormKernel` (Kahan + Newton-Raphson) produces the SAME wrong result — so the bug is NOT in FP32 accumulation precision or rsqrt approximation. The PTX kernel is fundamentally miscomputing on sm_121.

### Connection to albor/entrenar#309

The albor project found entrenar training converges 21x slower than PyTorch. If the RMSNorm warp shuffle reduction produces wrong values on sm_121, ALL gradient norms are also wrong (RMSNorm is used in every layer's backward pass too). This would explain the training quality gap.

### Fix Strategy

1. The warp shuffle `shfl.sync.down.b32` on sm_121 may have a hardware bug  
2. Or: the JIT-compiled reduction code has a wrong reduction tree
3. Test: replace shuffle reduction with shared memory reduction as a workaround
4. Test: dump the raw partial sums from each warp lane to find which lane is wrong
